### ISSUE TITLE
[MDEP-204] go-offline fails to resolve artifact available in maven reactor

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-dependency-plugin#1186]
**[Stevo Slavić](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sslavic)** opened **[MDEP-204](https://issues.apache.org/jira/browse/MDEP-204?redirect=false)** and commented

Attached is example project, for which IMO dependency:go-offline should be able to resolve all dependencies as they are only intermodule dependencies within same multimodule project which haven't been installed yet but are available in maven reactor so can be considered as resolvable in offline mode - instead dependency:go-offline fails to resolve these dependencies.

---

**Affects:** 2.0, 2.1

**Attachments:**
- [mvn-dependency-go-offline-failing-example.zip](https://issues.apache.org/jira/secure/attachment/12715169/mvn-dependency-go-offline-failing-example.zip) (_5.20 kB_)

**Issue Links:**
- [MDEP-516](https://issues.apache.org/jira/browse/MDEP-516) Go-offline does not find module dependencies in multi-module build.
 (_**"is duplicated by"**_)
- [MDEP-409](https://issues.apache.org/jira/browse/MDEP-409) dependency:tree fails for a multi-module project where one module depends on another and "mvn install" has not been run yet

**Remote Links:**
- [GitHub Pull Request #20
](https://github.com/apache/maven-dependency-plugin/pull/20)
- [GitHub Pull Request #21
](https://github.com/apache/maven-dependency-plugin/pull/21)

13 votes, 26 watchers

**Comment by jira-importer** [comment]
**[Louis Lecaroz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=llecaroz)** commented

Having the same issue on a project in Maven 2.2.1.
I am also blocked on a maven deployment on some remote dev workstation due to this error.... developers needs to go offline with a big project containing some child modules  having dependencies exactly like in this example attached in this issue to work offline & remotly & the go-offline fails with the same error.

I was able to reproduce this issue with the attached zip above also in 2.2.1

**Comment by jira-importer** [comment]
**[manuel aldana](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=aldana)** commented

The same for me here (maven 2.1.0), I think this is a severe issue. The only workaround to use dependency:resolve/tree etc. is to do a **mvn install** beforehand, so the modules are inside repository.

**Comment by jira-importer** [comment]
**[jonathan gold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jgold.bg)** commented

I'd also like to request that the plugin be changed to be reactor aware. I have projects with intermodule dependencies, and it appears that the only way I can use the dependency:tree plugin is to first install my project into my local repo. Full context of the discussion is here: http://markmail.org/message/z2sya75pynr3ia2k?q=list:org%2Eapache%2Emaven%2Eusers

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

patches welcome ;-)

**Comment by jira-importer** [comment]
**[Corneil du Plessis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=corneil.duplessis@gmail.com)** commented

I'm amazed that this issue is still unresolved. Does no one use dependency:go-offline for multi-module projects?

**Comment by jira-importer** [comment]
**[Bill Burcham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=propeller)** commented

An important use-case that didn't exist when this ticket was created is Docker.

A Dockerized project typically has a Dockerfile at the root of the project. The Dockerfile contains instructions for building an image. In the Dockerfile, it is useful to separate two kinds of processing:

1. downloading external dependencies
2. compiling (local) sources

This is the pattern for Dockerfiles in Python, Ruby, Node/JavaScript and others. The value of structuring the Dockerfile this way is that after building an initial image, subsequent source code changes do not incur the work in (1). This shortens the edit-compile-run loop.

It might seem that mvn dependency:go-offline is ideally suited to (1) except for the fact that it fails when a local project (A) depends on another local project (B) if B has not yet been compiled and installed. But B cannot be installed until after step (2).

Indeed, what is the point of go-offline, if it only (in general) succeeds after install? If mvn install succeeds, all external dependencies will have been cached locally, obviating the need to run mvn dependency:go-offline at all. Right?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Still fails with 3.0.0.

**Comment by jira-importer** [comment]
**[Yevhen Tienkaiev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hronom)** commented

Still not work on 3.0.2, it not allows to run in CircleCI service, wtf?

**Comment by jira-importer** [comment]
**[Hauke Jäger](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yegair)** commented

I just stumbled on this issue which is still present in 3.6.0 and I honestly have to say i can't believe this is still open after almost 10 years of time to fix it.

We are living in the times of Docker multi stage builds but Maven makes it very hard to being used there.

I will have to go for Gradle...

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Hauke Jäger](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yegair), granted. That's pretty sad from your POV. Consider that this is a volunteer project, none of us gets paid for this.

**Comment by jira-importer** [comment]
**[Jakub Bochenski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbochenski)** commented

You might be interested in https://github.com/qaware/go-offline-maven-plugin

**Comment by jira-importer** [comment]
**[Hauke Jäger](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yegair)** commented

[Jakub Bochenski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbochenski), thank you a lot. I will definitely give it a try.

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) you're absolutely right. I shouldn't be complaining and instead solve it myself and maybe contribute it, if I get it to work.

**Comment by jira-importer** [comment]
**[Maarten Mulders](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mthmulders)** commented

It seems that when the project is not going to be packaged, nor going to be compiled, the `org.apache.maven.ReactorReader` considers the project output dir to be non-present.

Because of this, the `org.apache.maven.ReactorReader` considers the requested artifact as not part of the Reactor. The `org.apache.maven.repository.internal.DefaultVersionResolver` that invoked it then decides to contact the remote repositories to see if they have the artifact.

As a consequence, doing `mvn dependency:go-offline compile` succeeds - but that's not really a solution :-)

**Comment by jira-importer** [comment]
**[Maarten Mulders](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mthmulders)** commented

I've shortly discussed this with [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) and we've analysed the situation. It seems to us that we cannot use the default Maven dependency resolution mechanism here, as it doesn't allow to exclude certain artifacts (for example, those we _know_ are in the Reactor).

I would propose to rewrite this Mojo so it roughly does the following:
* analyse the dependencies for the current module
* see if any of those are in the Reactor already
* attempt to resolve all dependencies that are _not_ in the reactor.

Regarding the latter, I think we could do it by following the logic in `MojoExecutor`'s `ensureDependenciesAreResolved` method, except when we invoke the `LifeCycleDependencyResolver` we supply the artifacts that are already in the Reactor.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I would expect that the code will be very similar to the `ResolvePluginsMojo`. The have the some purpose, but just a different set of artifacts.

**Comment by jira-importer** [comment]
**[Maarten Mulders](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mthmulders)** commented

I've made an attempt to rewrite this goal. It passes existing integration tests, and the one that I've added for this issue. There's a PR [at GitHub](https://github.com/apache/maven-dependency-plugin/pull/20).

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [23b7ca8790ae14175ed8e3a20c75c6274efe5ad8](https://gitbox.apache.org/repos/asf?p=maven-dependency-plugin.git;a=commit;h=23b7ca8790ae14175ed8e3a20c75c6274efe5ad8)
Thanks for the patch!
